### ISSUE TITLE
Is there anyway that I get the predicted transformation of an image?

### ISSUE THREAD

**Issue by petercmh01** [wpeebles/gangealing#35]
Hi: 
I'm quite new from the STN model. Just like the title says, for every image that I align by using a trained STN, I want to get each of their predicted transformation (i.e., the predicted set of transformation matrices). Is there anyway I can extract that from the STN model ? 

For some reason, I'm trying to use the predicted transformation of an image to apply on another different image. I saw in the forward function of STN there are return_flow, return_wrap function that can produce matrix, but not sure which should I take.

Thanks in advance : )

**Comment by wpeebles** [comment]
Hi @petercmh01, yeah you can do this. The behavior of `return_flow` and `return_warp` depends a bit on the specific STN you're using. If you're using a similarity --> flow `ComposedSTN` (this applies to all of our pre-trained models), `return_warp=True` will return a `(N,H,W,2)` sampling grid representing the full transformation (composing similarity and flow), and `return_flow=True` will return a `(N,H,W,2)` tensor representing the residual flow field predicted by the flow STN (contains no information about the similarity transformation). 

If you're using a similarity-only STN, `return_warp=True` will still return a sampling grid, but `return_flow=True` will return the `(N,2,3)` affine matrix representing the predicted similarity transformation.

There's also a third option for `ComposedSTN`,`return_intermediates=True`, which will return a list of the intermediate warped images and a list of the corresponding `(N,H,W,2)` sampling grids, where the lists are each of size equal to the number of STNs (two in the case of our pre-trained models). `return_intermediates` is useful if you want the sampling grid for both the similarity transformation as well as the full similarity + flow transformation. 

Let me know if this wasn't clear or you have other questions

**Comment by petercmh01** [comment]
thanks so much!
Just to double check, if I load my t_ema and called t_ema.forward(image_tensor, return_warp=True), I'll get both similarity and flow as full transformation in my output, is that correct? 
I'm actually also having a little issue just input an image into the composed stn. After I load t_ema (which I assume I load it correct?) and called t_ema(image_tensor) or t_ema.forward(image_tensor), I got some matrix multiply shape error in one of the linear layer. I wonder what size / shape of image tensor should I use to input into the model? 

It seems like i had no issue when use your default script of propagate_to_image or mix_reality on normal inferences task

**Comment by wpeebles** [comment]
Yeah usage is like this:

```python
x_aligned, sampling_grid = t_ema(image_tensor, return_warp=True)
```
Regarding the input image to the STN, it (currently) needs to be square, and it needs to have a resolution that is a multiple of `flow_size` (default value of `flow_size` is 128, so e.g., 128x128, 256x256, 512x512, etc. are acceptable input resolutions). Whatever the resolution is, you should pass it via the argument `supersize` when instantiating the STN.

**Comment by petercmh01** [comment]
I tried image_tensor using (1,3,256,256) and (3,256,256) but both did not work, with this following error: 

-------------------------------------------------------------------------
```

RuntimeError                              Traceback (most recent call last)
[<ipython-input-103-5bf4ad1c7d4a>](https://localhost:8080/#) in <module>
----> 1 t_ema(img.cuda())

7 frames
[/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in _call_impl(self, *input, **kwargs)
   1100         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1101                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1102             return forward_call(*input, **kwargs)
   1103         # Do not call functions when jit is used
   1104         full_backward_hooks, non_full_backward_hooks = [], []

[/content/gangealing/models/spatial_transformers/spatial_transformer.py](https://localhost:8080/#) in forward(self, input_img, return_warp, return_flow, return_sim, return_intermediates, output_resolution, unfold, iters, alpha, warp_policy, input_img_for_sampling, **stn_forward_kwargs)
    114                           input_img_for_sampling=source_pixels, base_warp=warp,
    115                           output_resolution=output_resolution_t, unfold=unfold_t,
--> 116                           iters=iters_t, alpha=alpha_t, warp_policy=warp_policy_t, **stn_forward_kwargs)
    117             out, grid, flow_or_matrix = stn_out
    118             if self.num_heads > 1 and warp_policy == 'cartesian' and i == 0:

[/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in _call_impl(self, *input, **kwargs)
   1100         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1101                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1102             return forward_call(*input, **kwargs)
   1103         # Do not call functions when jit is used
   1104         full_backward_hooks, non_full_backward_hooks = [], []

[/content/gangealing/models/spatial_transformers/spatial_transformer.py](https://localhost:8080/#) in forward(self, input_img, output_resolution, iters, return_warp, return_flow, return_intermediates, return_out_of_bounds, intermediate_output_resolution, stop_grad, alpha, padding_mode, input_img_for_sampling, image_bounds, warp_policy, unfold, base_warp)
    510                                        padding_mode=padding_mode, input_img_for_sampling=input_img_for_sampling,
    511                                        return_out_of_bounds=return_out_of_bounds, image_bounds=image_bounds,
--> 512                                        warp_policy=warp_policy, unfold=unfold, base_warp=base_warp)
    513         else:  # Continually apply the STN to its own output, composing the predicted warps at each iteration:
    514             return self.iterated_forward(input_img, output_resolution=output_resolution, iters=iters,

[/content/gangealing/models/spatial_transformers/spatial_transformer.py](https://localhost:8080/#) in single_forward(self, input_img, output_resolution, return_warp, return_flow, return_out_of_bounds, base_warp, input_img_for_sampling, stop_grad, alpha, padding_mode, image_bounds, warp_policy, unfold, pack)
    594         if not self.is_flow:
    595             out = out.view(batch, -1)
--> 596             out = self.final_linear(out)
    597 
    598         output_resolution = output_resolution if output_resolution is not None else self.stn_in_size

[/usr/local/lib/python3.7/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in _call_impl(self, *input, **kwargs)
   1100         if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1101                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1102             return forward_call(*input, **kwargs)
   1103         # Do not call functions when jit is used
   1104         full_backward_hooks, non_full_backward_hooks = [], []

[/content/gangealing/models/stylegan2/networks.py](https://localhost:8080/#) in forward(self, input)
    146     def forward(self, input):
    147         if self.activation:
--> 148             out = F.linear(input, self.weight * self.scale)
    149             out = fused_leaky_relu(out, self.bias * self.lr_mul)
    150 

[/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py](https://localhost:8080/#) in linear(input, weight, bias)
   1846     if has_torch_function_variadic(input, weight, bias):
   1847         return handle_torch_function(linear, (input, weight, bias), input, weight, bias=bias)
-> 1848     return torch._C._nn.linear(input, weight, bias)
   1849 
   1850 

RuntimeError: mat1 and mat2 shapes cannot be multiplied (1x6144 and 8192x512)
```

----------------------------------

When I load the model I used: 

```
parser = base_eval_argparse()
args = parser.parse_args()
t_ema, classifier = load_stn(args, load_classifier=True)
```

Just like how mix_reality did. 

The special thing about my model was I had dim_latent=512, but when I run mix_reality I didn't recall I had to specify that and it run with no error. Any ideas?

**Comment by petercmh01** [comment]
thank you so much, I have got my desired sampling_grid! I'll experiment with it for a while!

**Comment by petercmh01** [comment]
looks like I got many very huge number (either very huge positive or very huge negative) in the grid. Is there a normalization required?

**Comment by wpeebles** [comment]
Depends what you're using the sampling grid for. If you just want to use `F.grid_sample` you shouldn't need to normalize it, but note that the input image to the STN does need to be normalized to [-1, 1]. You can verify the sampling grid is correct by running:

```python
import torch.nn.functional as F
x_aligned = F.grid_sample(x, sampling_grid, "bilinear", "border")
```
This should give approximately the same image as the one directly returned by the STN (to get the exact same result, you need to use the [mipmap warper](https://github.com/wpeebles/gangealing/blob/main/models/spatial_transformers/antialiased_sampling.py) instead of `F.grid_sample`).

**Comment by petercmh01** [comment]
does congeal_point works the same way? if I called congeal_point(img_tensor, points, return_full=True), does wrapA output give me similar things to sampling grid?

**Comment by petercmh01** [comment]
This is a concept question rather than a model specific question. Thanks really so so much for really helping me with so many questions 

In here 
https://github.com/wpeebles/gangealing/blob/ffa6387c7ffd3f7de76bdc693dc2272e274e9bfd/models/spatial_transformers/spatial_transformer.py#L151

If instead of looking for pointsB, I now have pointsB and grid (which is what we discussed to get earlier in this issue), but not points_congealed, is there anyway that I can rewrite this line and get point_congealed?

In other words, is there anyway that I can use my grid and uncongealed coodinate to find the coordinates after congealed?

update: i just solve this issue by search the grid location
